overlord/configstate/configcore/kernel.go

```diff
@@ -0,0 +1,213 @@
+		params, err := osutil.KernelCommandLineSplit(cmdAppend)
+		if err != nil {
+			return err
+		}
+		if opt == optionKernelCmdlineAppend {
+			// check against allowed values from gadget
+			if err := validateCmdlineParamsAreAllowed(c.State(), devCtx, params); err != nil {
+				return err
+			}
+		} else { // OptionKernelDangerousCmdlineAppend
+			if devCtx.Model().Grade() != asserts.ModelDangerous {
+				// TODO we should return an error if this is an API call
+				// and do nothing if setting defaults (so gadget can be
+				// reused with different models).
+				logger.Noticef("WARNING: %s ignored as this is not a dangerous model", opt)
+			}
+		}
+	}
+
+	return nil
+}
+
+func createApplyCmdlineChange(c RunTransaction, kernelOpts []string) (*state.Change, error) {
+	st := c.State()
+	st.Lock()
+	defer st.Unlock()
+
+	// error out if some other change is touching the kernel command line
+	if err := snapstate.CheckUpdateKernelCommandLineConflict(st, ""); err != nil {
+		return nil, err
+	}
+
+	// precalculate task arguments, so we do not need to destroy change/task
+	// if there is an error
+	args := []struct {
+		name    string
+		cmdline string
+	}{}
+	for _, opt := range kernelOpts {
+		cmdline, err := coreCfg(c, opt)
+		if err != nil {
+			return nil, err
+		}
+		// take just *cmdline-append (it should not fail)
+		name := strings.Split(opt, ".")[2]
+		args = append(args, struct {
+			name    string
+			cmdline string
+		}{
+			name: name, cmdline: cmdline})
+	}
+
+	// We need to create a new change that will change the kernel
+	// command line and wait for it to finish, otherwise we cannot
+	// wait on the changes to happen.
+	// TODO fix this in the future.
+	cmdlineChg := st.NewChange("apply-cmdline-append",
+		i18n.G("Update kernel command line due to change in system configuration"))
+	// Add task to the new change to set the new kernel command line
+	t := st.NewTask("update-gadget-cmdline",
+		"Update kernel command line from system configuration")
+	// Pass options to the task (changes in the options are not
+	// committed yet so the task cannot simply get them from the
+	// configuration)
+	for _, arg := range args {
+		t.Set(arg.name, arg.cmdline)
+	}
+
+	cmdlineChg.AddTask(t)
+	st.EnsureBefore(0)
+
+	return cmdlineChg, nil
+}
+
+func isDangerousModel(st *state.State) (bool, error) {
+	st.Lock()
+	defer st.Unlock()
+	devCtx, err := devicestate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return false, err
+	}
+
+	return devCtx.Model().Grade() == asserts.ModelDangerous, nil
+}
+
+func handleCmdlineAppend(c RunTransaction, opts *fsOnlyContext) error {
+	kernelOpts := changedKernelConfigs(c)
+	if len(kernelOpts) == 0 {
+		return nil
+	}
+	logger.Debugf("handling %v", kernelOpts)
+
+	st := c.State()
+	isDangModel, err := isDangerousModel(st)
+	if err != nil {
+		return err
+	}
+	// nothing to do if non-dangerous model and the only option set is
+	// the dangerous one, we simply return with success
```

Comment:
Does this need a similar TODO as we did in line 109 (https://github.com/snapcore/snapd/pull/12560/files#diff-c4ea4e76e75ed60a40fba6f9e5a866b5dcecd9605707deeba787d7d8c6206ef3R105) ?

Reply:
I'm not sure, validate will be done always after the handler is called, so checking in the former should be enough afaiu, but maybe I am missing something.

Reply:
Aha, I think I misread, sorry. I think this is fine as is. My confusion.